pkg/apis/provisioning/v1alpha5/requirements.go

```diff
@@ -70,126 +76,213 @@ var (
 		if WellKnownLabels.Has(requirement.Key) {
-			requirements = requirements.Add(requirement)
+			requirements = append(requirements, requirement)
 		}
 	}
-	return requirements
+	return NewRequirements(requirements...)
+}
+
+// Add function returns a new Requirements object with new requirements inserted.
+func (r Requirements) Add(requirements ...v1.NodeSelectorRequirement) Requirements {
+	// Deep copy to avoid mutating existing requirements
+	r = *r.DeepCopy()
+	if r.requirements == nil {
+		r.requirements = map[string]sets.Set{}
+	}
+	for _, requirement := range requirements {
+		if normalized, ok := NormalizedLabels[requirement.Key]; ok {
+			requirement.Key = normalized
+		}
+		r.Requirements = append(r.Requirements, requirement)
+		switch requirement.Operator {
+		case v1.NodeSelectorOpIn:
+			r.requirements[requirement.Key] = r.Values(requirement.Key).Intersection(sets.NewSet(requirement.Values...))
+		case v1.NodeSelectorOpNotIn:
+			r.requirements[requirement.Key] = r.Values(requirement.Key).Intersection(sets.NewComplementSet(requirement.Values...))
+		}
+	}
+	return r
 }
 
 // Keys returns unique set of the label keys from the requirements
-func (r Requirements) Keys() []string {
-	keys := sets.NewString()
-	for _, requirement := range r {
+func (r Requirements) Keys() stringsets.String {
+	keys := stringsets.NewString()
+	for _, requirement := range r.Requirements {
 		keys.Insert(requirement.Key)
 	}
-	return keys.UnsortedList()
+	return keys
+}
+
+// Values returns the sets of values allowed by all included requirements
+// following a denylist method. Values are allowed except specified
+func (r Requirements) Values(key string) sets.Set {
+	if _, ok := r.requirements[key]; !ok {
+		return sets.NewComplementSet()
+	}
+	return r.requirements[key]
+}
+
+func (r Requirements) Zones() stringsets.String {
+	return r.Values(v1.LabelTopologyZone).Values()
+}
+
+func (r Requirements) InstanceTypes() stringsets.String {
+	return r.Values(v1.LabelInstanceTypeStable).Values()
+}
+
+func (r Requirements) Architectures() stringsets.String {
+	return r.Values(v1.LabelArchStable).Values()
+}
+
+func (r Requirements) OperatingSystems() stringsets.String {
+	return r.Values(v1.LabelOSStable).Values()
+}
+
+func (r Requirements) CapacityTypes() stringsets.String {
+	return r.Values(LabelCapacityType).Values()
+}
+
+// Validate validates the feasibility of the requirements.
+func (r Requirements) Validate() (errs *apis.FieldError) {
+	for i, requirement := range r.Requirements {
+		for _, err := range validation.IsQualifiedName(requirement.Key) {
+			errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("%s, %s", requirement.Key, err), "key"))
+		}
+		for _, value := range requirement.Values {
+			for _, err := range validation.IsValidLabelValue(value) {
+				errs = errs.Also(apis.ErrInvalidArrayValue(fmt.Sprintf("%s, %s", value, err), "values", i))
+			}
+		}
+		if !functional.ContainsString(SupportedNodeSelectorOps, string(requirement.Operator)) {
```

Comment:
Can we make this set.String{}.Has()

Reply:
We can also deprecate functional.IntersectStringSlice

Reply:
done